test-data/unit/check-selftype.test

```diff
@@ -468,16 +468,225 @@ class B(A[Q]):
+
+class Sub(Base[List[int]]): ...
+class BadSub(Base[int]): ...
+
+reveal_type(Sub().get_item())  # N: Revealed type is 'builtins.int'
+BadSub().get_item()  # E: Invalid self argument "BadSub" to attribute function "get_item" with type "Callable[[Base[List[S]]], S]"
+[builtins fixtures/list.pyi]
+
+[case testMixinAllowedWithProtocol]
+from typing import TypeVar
+from typing_extensions import Protocol
+
+class Resource(Protocol):
+    def close(self) -> int: ...
+
+class AtomicClose:
+    def atomic_close(self: Resource) -> int:
+        return self.close()
+
+T = TypeVar('T', bound=Resource)
+class Copyable:
+    def copy(self: T) -> T: ...
+
+class File(AtomicClose, Copyable):
+    def close(self) -> int:
+       ...
+
+class Bad(AtomicClose, Copyable):
+    ...
+
+f: File
+b: Bad
+f.atomic_close()  # OK
+b.atomic_close()  # E: Invalid self argument "Bad" to attribute function "atomic_close" with type "Callable[[Resource], int]"
+
+reveal_type(f.copy())  # N: Revealed type is '__main__.File*'
+b.copy()  # E: Invalid self argument "Bad" to attribute function "copy" with type "Callable[[T], T]"
+
+[case testBadClassLevelDecoratorHack]
+from typing_extensions import Protocol
+from typing import TypeVar, Any
+
+class FuncLike(Protocol):
+    __call__: Any
+F = TypeVar('F', bound=FuncLike)
+
+class Test:
+    def _deco(func: F) -> F: ...
+
+    @_deco
+    def meth(self, x: str) -> int: ...
+
+reveal_type(Test().meth)  # N: Revealed type is 'def (x: builtins.str) -> builtins.int'
+Test()._deco  # E: Invalid self argument "Test" to attribute function "_deco" with type "Callable[[F], F]"
+
+[case testSelfTypeTrickyExample]
+from typing import *
+
+In = TypeVar('In')
+Out = TypeVar('Out')
+Mid = TypeVar('Mid')
+NewOut = TypeVar('NewOut')
+
+class Lnk(Generic[In, Out]):
+    def test(self: Lnk[In, Mid], other: Lnk[Mid, NewOut]) -> Lnk[In, NewOut]: ...
+
+class X: pass
+class Y: pass
+class Z: pass
+
+a: Lnk[X, Y] = Lnk()
+b: Lnk[Y, Z] = Lnk()
+
+a.test(b)
+b.test(a)  # E: Argument 1 to "test" of "Lnk" has incompatible type "Lnk[X, Y]"; expected "Lnk[Z, Y]"
+
+[case testSelfTypeReallyTrickyExample]
+from typing import *
+
+In = TypeVar('In')
+Out = TypeVar('Out')
+Other = TypeVar('Other')
+
+_1 = TypeVar('_1')
+_2 = TypeVar('_2')
+__1 = TypeVar('__1')
+__2 = TypeVar('__2')
+
+class Lnk(Generic[In, Out]):
+    @overload
+    def __rshift__(self, other: Lnk[Out, Other]) -> Lnk[In,Other]: ...
+    @overload
+    def __rshift__(self: Lnk[In, Tuple[_1, _2]],
+                   other: Tuple[Lnk[_1, __1], Lnk[_2, __2]]) -> Lnk[In, Tuple[__1, __2]]: ...
+    def __rshift__(self: Any, other: Any) -> Any:
+        ...
+
+a: Lnk[str, Tuple[str, int]] = Lnk()
+b: Lnk[str, int] = Lnk()
+c: Lnk[int, float] = Lnk()
+
+d: Lnk[str, float] = b >> c  # OK
+e: Lnk[str, Tuple[int, float]] = a >> (b, c)  # OK
+f: Lnk[str, Tuple[float, int]] = a >> (c, b) # E: Unsupported operand types for >> ("Lnk[str, Tuple[str, int]]" and "Tuple[Lnk[int, float], Lnk[str, int]]")
+
```

Comment:
I'm sure this is already handled, but maybe it could be worth checking to make sure we correctly report errors with custom self-types within the actual methods?

For example, maybe we could add a test case where we have a class `Foo[T]` that defines two methods `f1(self: Foo[int]) -> None` and `f2(self: Foo[str]) -> None`. Then, try calling f2 from inside f1 and make sure that reports an error?